### DISCUSSION TITLE
Does bpftrace read the cpp library or build with clang++ ?

### DISCUSSION THREAD

**Discussion by Jackyen** [bpftrace/bpftrace#2311] [Q&A]
Hi all, 
I want to probe a process and read the data struct for calculating the real address.

I included the library with -I  xxx/included/, ..., but I still get something wrong.
I try to build with clang and clang++, but only the clang++ work.

Does bpftrace can support the CPP library or build with clang++?

**Comment by fbs** [comment]
Hi,

Can you give a concrete example and error output? bpftrace doesn't have any c++ specific features except for demangling of symbols and dwarf support

**Comment by Jackyen** [comment]
For example: 
I want to include one of the library from my processor. 

```
## simple.bt
#include "tscore/ts_file.h"
uprobe:/opt/edge/trafficserver/9.1/bin/traffic_server:"HttpTransact::HandleRequest" {
    printf("Hi\n");
}
```
```
[]# bpftrace --unsafe  -I /usr/local/include -I /usr/include -I /trafficserver/include simple.bt
/trafficserver/include/tscore/ts_file.h:26:10: fatal error: 'string' file not found
```

I doubt it is related to the build way.
I try to build with clang and clang++,  clang will show the error and clang++ will pass.

**Reply by dxuuu** [reply]
bpftrace does not support parsing C++ header files. Only C. 

If you're trying to read a member from a class, you might find dwarf support useful. See the section on dwarf support here: https://github.com/iovisor/bpftrace/blob/master/docs/reference_guide.md#4-uprobeuretprobe-dynamic-tracing-user-level-arguments
